src/trio/_tools/sync_requirements.py

```diff
@@ -0,0 +1,103 @@
+#!/usr/bin/env python3
+
+"""Sync Requirements - Automatically upgrade test requirements pinned
+versions from pre-commit config file."""
+
+from __future__ import annotations
+
+import sys
+from pathlib import Path
+from typing import TYPE_CHECKING
+
+from yaml import load as load_yaml
+
+if TYPE_CHECKING:
+    from collections.abc import Generator
+
+    from yaml import CLoader as _CLoader, Loader as _Loader
+
+    Loader: type[_CLoader | _Loader]
+
+try:
+    from yaml import CLoader as Loader
+except ImportError:
+    from yaml import Loader
+
+
+def yield_pre_commit_version_data(
+    pre_commit_text: str,
+) -> Generator[tuple[str, str], None, None]:
+    """Yield (name, rev) tuples from pre-commit config file."""
+    pre_commit_config = load_yaml(pre_commit_text, Loader)
+    for repo in pre_commit_config["repos"]:
+        if "repo" not in repo or "rev" not in repo:
+            continue
+        url = repo["repo"]
+        name = url.rsplit("/", 1)[-1]
+        rev = repo["rev"].removeprefix("v")
+        yield name, rev
+
+
+def update_requirements(
+    requirements: Path,
+    version_data: dict[str, str],
+) -> bool:
+    """Return if updated requirements file.
+
+    Update requirements file to match versions in version_data."""
+    changed = False
+    old_lines = requirements.read_text(encoding="utf-8").splitlines(True)
+
+    with requirements.open("w", encoding="utf-8") as file:
+        for line in old_lines:
+            # If comment or not version mark line, ignore.
+            if line.startswith("#") or "==" not in line:
+                file.write(line)
+                continue
+            name, rest = line.split("==", 1)
+            # Maintain extra markers if they exist
+            old_version = rest.strip()
+            extra = "\n"
+            if " " in rest:
+                old_version, extra = rest.split(" ", 1)
+                extra = " " + extra
+            version = version_data.get(name)
+            # If does not exist, skip
+            if version is None:
+                file.write(line)
+                continue
+            # Otherwise might have changed
+            new_line = f"{name}=={version}{extra}"
+            if new_line != line:
+                if not changed:
+                    changed = True
+                    print("Changed test requirements version to match pre-commit")
+                print(f"{name}=={old_version} -> {name}=={version}")
+            file.write(new_line)
+    return changed
```

Comment:
This should also check whether every entry in version_data was used

Reply:
I don't immediately understand why this should be the case. For example, `typos` exists only as a pre-commit hook, and in fact I don't know if it even exists as a python package. Similar with `zizmor`.

Reply:
Yeah this would only make sense if https://github.com/python-trio/trio/pull/3234#discussion_r2018183152 was implemented.